src/common/imageio_jpegxl.c

```diff
@@ -0,0 +1,242 @@
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+
+  // Grand Decoding Loop
+  while(1)
+  {
+    status = JxlDecoderProcessInput(decoder);
+
+    if(status == JXL_DEC_ERROR)
+    {
+      fprintf(stderr, "[jpegxl_open] ERROR: JXL decoding failed\n");
+      JxlResizableParallelRunnerDestroy(runner);
+      JxlDecoderDestroy(decoder);
+      free(read_buffer);
+      return DT_IMAGEIO_FILE_CORRUPTED;
+    }
+
+    if(status == JXL_DEC_NEED_MORE_INPUT)
+    {
+      fprintf(stderr, "[jpegxl_open] ERROR: JXL data incomplete\n");
+      JxlResizableParallelRunnerDestroy(runner);
+      JxlDecoderDestroy(decoder);
+      free(read_buffer);
+      return DT_IMAGEIO_FILE_CORRUPTED;
+    }
+
+    if(status == JXL_DEC_BASIC_INFO)
+    {
+      if(JxlDecoderGetBasicInfo(decoder, &basicinfo) != JXL_DEC_SUCCESS)
+      {
+        fprintf(stderr, "[jpegxl_open] ERROR: JXL basic info not available\n");
+        JxlResizableParallelRunnerDestroy(runner);
+        JxlDecoderDestroy(decoder);
+        free(read_buffer);
+        return DT_IMAGEIO_FILE_CORRUPTED;
+      }
+
+      // Unlikely to happen, but let there be a sanity check
+      if(basicinfo.xsize == 0 || basicinfo.ysize == 0)
+      {
+        fprintf(stderr,"[jpegxl_open] ERROR: JXL image declares zero dimensions\n");
+        JxlResizableParallelRunnerDestroy(runner);
+        JxlDecoderDestroy(decoder);
+        free(read_buffer);
+        return DT_IMAGEIO_FILE_CORRUPTED;
+      }
+
+
+      num_threads = JxlResizableParallelRunnerSuggestThreads(basicinfo.xsize, basicinfo.ysize);
+      JxlResizableParallelRunnerSetThreads(runner, num_threads);
+
+      continue;    // go to next loop iteration to process rest of the input
+    }
+
+    if(status == JXL_DEC_COLOR_ENCODING)
+    {
+      if(JxlDecoderGetICCProfileSize(decoder, &pixel_format, JXL_COLOR_PROFILE_TARGET_DATA, &icc_size) == JXL_DEC_SUCCESS)
+      {
+        if(icc_size)
+        {
+          img->profile_size = icc_size;
+          img->profile = (uint8_t *)g_malloc0(icc_size);
+          JxlDecoderGetColorAsICCProfile(decoder, &pixel_format, JXL_COLOR_PROFILE_TARGET_DATA, img->profile, icc_size);
+        }
+      } else
+        img->colorspace = DT_IMAGE_COLORSPACE_SRGB;
+    continue;    // go to next iteration to process rest of the input
+    }
+
+    if(status == JXL_DEC_NEED_IMAGE_OUT_BUFFER)
+    {
+      img->width = basicinfo.xsize;
+      img->height = basicinfo.ysize;
+      img->buf_dsc.channels = 4;
+      img->buf_dsc.datatype = TYPE_FLOAT;
+      float *mipbuf = (float *)dt_mipmap_cache_alloc(mbuf, img);
+      if(!mipbuf)
+      {
+        JxlResizableParallelRunnerDestroy(runner);
+        JxlDecoderDestroy(decoder);
+        g_free(read_buffer);
+        fprintf(stderr, "[jpegxl_open] could not alloc full buffer for image: %s\n", img->filename);
+        return DT_IMAGEIO_CACHE_FULL;
+      }
+      JxlDecoderSetImageOutBuffer(decoder, &pixel_format, mipbuf, basicinfo.xsize * basicinfo.ysize * 4 * 4);
```

Comment:
Also, do we perhaps want to set [`JxlDecoderSetUnpremultiplyAlpha()`](https://github.com/libjxl/libjxl/blob/v0.7.x/lib/include/jxl/decode.h#L517-L533)?

Reply:
> We also need to [make sure the alpha channel is set to 0 again](https://github.com/darktable-org/darktable/pull/12560#discussion_r985058176)

In darktable the fourth channel is not an alpha channel, it is an unused channel. If you are certain (or suspect) that the lack of zeroing of this channel should cause problems, I think you can quickly find them and create a bug report? I didn't see such problems when working with (non-zeroed) webp , but maybe I was looking in the wrong place (although, say, I worked with masks and didn't notice anything wrong).

Reply:
> Also, do we perhaps want to set [`JxlDecoderSetUnpremultiplyAlpha()`](https://github.com/libjxl/libjxl/blob/v0.7.x/lib/include/jxl/decode.h#L517-L533) beforehand?

In any case, it is difficult to imagine that images with transparency will be imported into dt for further processing. It doesn't make sense. If the user will do this, then obviously only with the aim of using dt as a DAM for such images.

Reply:
> In darktable the fourth channel is not an alpha channel, it is an unused channel.

AFAIR this is used for creating masks, no?

Reply:
> > In darktable the fourth channel is not an alpha channel, it is an unused channel.
> 
> AFAIR this is used for creating masks, no?

Only the author of the corresponding code, or the developers who were involved in it at the time (reviewed, discussed the implementation, etc.) can say about this for sure. It is possible, for example, that this channel is indeed used for this purpose, but its content is not relied upon and the channel is filled with zeros before such use. I can only see it from a practical side, if we save time and don't fill the channel with zeroes on import will it cause bugs?

Now in front of me is a dt with two images, or rather the same image in two different formats, classic jpeg and webp, the loader of which does not fill channel 4 with zeros. I repeat the same steps in the retouch module with both images using masks. I don't see any differences.

Hence my suggestion: if someone can find the masking bug (I may not be looking hard enough), then adding channel filling with zeros that'll fix it is trivial. But adding this action, which is quite time-consuming for large images, just because of the fear that something might be wrong, does not seem worth it to me.

Reply:
> classic jpeg and webp, the loader of which does not fill channel 4 with zeros

But the jpeg loader does initialize the buffer w/ zeros, but granted does not reset the 4th channel to zeroes after reading from file which could potentially have the alpha channel (very rare for JPEGs I guess)...

> Hence my suggestion: if someone can find the masking bug (I may not be looking hard enough), then adding channel filling with zeros that'll fix it is trivial. But adding this action, which is quite time-consuming for large images, just because of the fear that something might be wrong, does not seem worth it to me.

Not my call to make unfortunately, although I am personally ok with this risk - as you said, only the original devs of the masking can decide on this w/ certainty.

Reply:
> In any case, it is difficult to imagine that images with transparency will be imported into dt for further processing. It doesn't make sense. If the user will do this, then obviously only with the aim of using dt as a DAM for such images.

While I don't think we can imagine all the use cases and should put such strict limits on users, I'm ok with putting this on hold until such a feature request comes in.